### PR TITLE
Support for Consul 1.4.0+ ACL system

### LINKED ISSUE #471: [Feature request] Provide support for Consul 1.4.0+ ACL System
With consul version 1.4.0 the ACL System changed : https://www.consul.io/docs/guides/acl-migrate-tokens.html
Even though the next major releases of Consul will still work with the older ACL system, it would be great to have this feature for this puppet module.

### PR DESCRIPTION
**Pull request by SkydiveMarius**
Resolves #471 

Example usage:
````puppet
consul_policy {'test_policy':
  description   => 'test description',
  rules         => [
      {
          'resource'    => 'service_prefix',
          'segment'     => 'test_service',
          'disposition' => 'read'
      },
      {
          'resource'    => 'key',
          'segment'     => 'test_key',
          'disposition' => 'write'
      },
  ],
  acl_api_token => 'e33653a6-0320-4a71-b3af-75f14578e3aa',
}

consul_token {'test_token':
  policies_by_name  => [
   'test_policy'
  ],
  policies_by_id    => [
    '652f27c9-d08d-412b-8985-9becc9c42fb2'
  ],
}
````

The current concept treats tokens and policies individually. A token can be assigned both policies managed by puppet (by_name) and external policies (by_id).
The new Consul system does not support any user specified Token ID anymore (s. https://github.com/hashicorp/consul/issues/4977), therefore the mapping is currently done by description.

Due to my limited Ruby or Puppet experience feedback and suggestions for improvements are highly welcome.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -199,6 +199,45 @@ consul to restart.
 
 ## ACL Definitions
 
+### Policy/Token system
+
+Starting with version 1.4.0, a new ACL system was introduces separating rules (policies) from tokens.
```
Comment (bastelfreak) [inline README.md]: yey doc updates \o/

File: lib/puppet/provider/consul_policy/default.rb [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    body.store('Description', policy.description)
+    body.store('Rules', policy.rules)
+
+    body
+  end
+
+  def delete_policy(id)
+    delete('/policy/' + id)
+  end
+end
```
Comment (bastelfreak) [inline lib/puppet/provider/consul_policy/default.rb]: Can you please add a newline at the end of this file?

File: lib/puppet/provider/consul_policy/default.rb [hunk 3]
```diff
@@ -0,0 +1,190 @@
+require 'json'
+require 'net/http'
+require 'pp'
```
Comment (bastelfreak) [inline lib/puppet/provider/consul_policy/default.rb]: Is `pp` actually required or just a leftover from debugging?

File: manifests/init.pp [hunk 4]
```diff
@@ -158,6 +179,13 @@
 #
 class consul (
   Hash $acls                                 = $consul::params::acls,
+  Hash $tokens                               = $consul::params::tokens,
```
Comment (bastelfreak) [inline manifests/init.pp]: can you enforce the datatypes even more? For hashes, you can define the datatypes for the keys and the values. For example:
```suggestion
  Hash[String[1],String[1]] $tokens          = $consul::params::tokens,
```

File: manifests/init.pp [hunk 5]
```diff
@@ -158,6 +179,13 @@
...
+  Hash $policies                             = $consul::params::policies,
+  String $acl_api_hostname                   = $consul::params::acl_api_hostname,
```
Comment (bastelfreak) [inline manifests/init.pp]: You can enforce the minimal string length like this:
```suggestion
  String[1] $acl_api_hostname                = $consul::params::acl_api_hostname,
```

Please also look through https://github.com/puppetlabs/puppetlabs-stdlib/tree/master/types, you might find mor strict datatypes that you could use.

File: manifests/init.pp [hunk 6]
```diff
@@ -158,6 +179,13 @@
...
+  String $acl_api_protocol                   = $consul::params::acl_api_protocol,
+  Integer $acl_api_port                      = $consul::params::acl_api_port,
```
Comment (bastelfreak) [inline manifests/init.pp]: Probably this type can be used https://github.com/puppetlabs/puppetlabs-stdlib/blob/master/types/port.pp ?
```suggestion
  Stdlib::Port $acl_api_port                 = $consul::params::acl_api_port,
```

**Comment by bastelfreak** [comment]
Hey @SkydiveMarius, thanks for the awesome PR! I made some small inline comments. Can you take a look at the current acceptance tests (https://github.com/solarkennedy/puppet-consul/blob/master/spec/acceptance/class_spec.rb) and extend them with tests for the new types/provider?

File: lib/puppet/provider/consul_policy/default.rb (hunk 3 above)
Comment (SkydiveMarius) [reply lib/puppet/provider/consul_policy/default.rb]: You are right, dependency has been removed.

File: manifests/init.pp (hunk 4 above)
Comment (SkydiveMarius) [reply manifests/init.pp]: Good idea. I tried to add some stricter types in the newest version.

File: manifests/init.pp (hunk 6 above)
Comment (SkydiveMarius) [reply manifests/init.pp]: Good idea from my point of view.
As far as I know this type was added in version 4.25 of stdlib. Is it ok to adjust the [current dependencies](https://github.com/solarkennedy/puppet-consul/blob/master/metadata.json#L13) accordingly?

**Comment by bastelfreak** [comment]
@SkydiveMarius thanks for the updates! Can you take a look at the failed travis job?

**Comment by SkydiveMarius** [comment]
Branch has been rebased, but some Travis builds are randomly failing with Timeout errors (ACL API requests). So far I was not able to reproduce this behavior locally.

In case of failure the content of /var/log/consul would probably help. Does rspec offer possibilities to display this information in case of an error?

**Comment by SkydiveMarius** [comment]
PR has been adjusted for the recent Consul 1.5.0 update.
As the module is now using the new accessor + secret ID predefinition feature, only Consul > 1.5.0 is supported.

File: manifests/init.pp [hunk 7]
```diff
@@ -157,50 +178,80 @@
...
-  Enum['stopped', 'running']     $service_ensure               = 'running',
-  Hash                           $services                     = {},
-  String[1]                      $user                         = $consul::params::user,
-  String[1]                      $version                      = '1.2.3',
-  Hash                           $watches                      = {},
-  Optional[String[1]]            $shell                        = $consul::params::shell,
-  Boolean                        $enable_beta_ui               = false,
-  Boolean                        $allow_binding_to_root_ports  = false,
+  Hash                           $acls                        = {},
+  Hash[String[1], Struct[{
```
Comment (bastelfreak) [inline manifests/init.pp]: can you please put those types into a custom datatype? That would make the file more readable.

Comment (SkydiveMarius) [reply manifests/init.pp]: Good idea. Replaced by type aliases.

**Comment by bastelfreak** [comment]
Hi @SkydiveMarius, thanks for the awesome work! @solarkennedy can you please take a look and merge it? In my opinion this is fine.

**Comment by solarkennedy** [comment]
looking...

File: lib/puppet/provider/consul_policy/default.rb [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+
+  def destroy
+    @property_flush[:ensure] = :absent
+  end
+
+  def flush
+    if @resource[:ensure] == :absent
+      if @existing_policy
+        @client.delete_policy(@existing_policy.id)
+        Puppet.notice(" Deleted Consul ACL policy #{@existing_policy.name} (ID: #{@existing_policy.id})")
```
Comment (solarkennedy) [inline lib/puppet/provider/consul_policy/default.rb]: nit: There is a leading space here in the log line

File: lib/puppet/provider/consul_policy/default.rb [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+      @resource[:id] = policy.id
+      Puppet.notice("Created Consul ACL policy #{policy.name} with ID #{policy.id}")
+    end
+
+    if @existing_policy && (@existing_policy.description != @resource[:description] || @existing_policy.rules != @rules_encoded)
+      @existing_policy.description = @resource[:description]
+      @existing_policy.rules = @rules_encoded
+
+      @client.update_policy(@existing_policy)
+      Puppet.notice(" Updated Consul ACL policy #{@existing_policy.name} (ID: #{@existing_policy.id})")
```
Comment (solarkennedy) [inline lib/puppet/provider/consul_policy/default.rb]: here too, intentional?

**Comment by solarkennedy** [comment]
This PR looks really solid @SkydiveMarius!
Tests, types, providers, etc.
It has been a long time since I wrote puppet code like this, but I'm a fan of getting the code out there and iterating. 

So I say we ship!

File: lib/puppet/provider/consul_policy/default.rb (hunk 8 above)
Comment (SkydiveMarius) [reply lib/puppet/provider/consul_policy/default.rb]: Thx, space has been removed.

File: lib/puppet/provider/consul_policy/default.rb (hunk 9 above)
Comment (SkydiveMarius) [reply lib/puppet/provider/consul_policy/default.rb]: By far not intended, space is removed.

**Comment by SkydiveMarius** [comment]
Thanks for the positive feedback!
Seems the current Travis fails have an unrelated reason:
````
curl: (22) The requested URL returned error: 404 Not Found
  	error: skipping http://yum.puppet.com/puppet6/puppet6-release-el-6.noarch.rpm - transfer failed
  	Retrieving http://yum.puppet.com/puppet6/puppet6-release-el-6.noarch.rpm
````

**Comment by solarkennedy** [comment]
Probably temporary. It really is 404, but I'll check back later in the day to make sure we have a green build and merge.

**Comment by bastelfreak** [comment]
I restarted the failed CI jobs

**Comment by bastelfreak** [comment]
The related issue for the broken mirror: https://groups.google.com/forum/#!topic/puppet-users/cCsGWKunBe4

**Comment by solarkennedy** [comment]
Build is green. Thanks again @SkydiveMarius for a really great first time contribution!

**Pull request merged**